umbraco/Umbraco-CMS
# V15: Implement sorting for the in-memory navigation structures (document and media)

## Details
- The `NavigationNode` model now keeps track of the `sortOrder` as well;
  - `Parent` and `Children` properties are refactored to reference `Guid` keys instead of the whole `NavigationNode` object - which means we need to resolve the `NavigationNode` from the key in all the properties' usages (i.e. **ContentNavigationServiceBase**).
- Adding a new `UpdateSortOrder()` to the `INavigationManagementService.cs`;
  - `Add()` changed its signature to have a `sortOrder` parameter as well, due to a special case when adding nodes directly to the root (where `parentKey` is null), the sort order must be provided to ensure the item appears in the correct position among other root-level items. We can't calculate the amount of children based on the parent here.
- Cache refreshers (document and media) are updated to update the sort order in the navigation structure;
- Added integration and unit tests for both document and media.

## Test
- Create a simple content type **with template**;
- Create a nested tree structure that you would use for sorting, something like:
```
Root
   - Child 1
     - Grandchild 1
     - Grandchild 2
   - Child 2
     - Grandchild 3
       - Great-grandchild 1
   - Child 3
     - Grandchild 4
```
- Create a new Partial View from "_List Descendants From Current Page_" and call it `ListDescendants.cshtml`;
- In the Root template use the following line `@await Html.PartialAsync("ListDescendants")` to display it, so the template looks like:
```cshtml
@using Umbraco.Cms.Web.Common.PublishedModels;
@inherits Umbraco.Cms.Web.Common.Views.UmbracoViewPage
@{
	Layout = null;
}

<h2>Descendants</h2>
@await Html.PartialAsync("ListDescendants")
```
- In the Content section, sort the children on a parent node;
- Verify that the tree from the template reflects those changes.
